Mark CXX module initializer with PACBTI attributes

vhscampos opened the pull request:
The CXX module initializer function, which is called at program startup, needs to be tagged with Pointer Authentication and Branch Target Identification marks whenever relevant.

Before this patch, in CPUs set up for PACBTI execution, the function wasn't protected with return address signing and no BTI instruction was inserted at the start of it, thus leading to an execution fault.

This patch fixes the issue by marking the function with the function attributes related to PAC and BTI if relevant.

vhscampos committed:
Mark CXX module initializer with PACBTI attributes

The CXX module initializer function, which is called at program startup,
needs to be tagged with Pointer Authentication and Branch Target
Identification marks whenever relevant.

Before this patch, in CPUs set up for PACBTI execution, the function
wasn't protected with return address signing and no BTI instruction was
inserted at the start of it, thus leading to an execution fault.

This patch fixes the issue by marking the function with the function
attributes related to PAC and BTI if relevant.

efriedma-quic commented on clang/lib/CodeGen/CGDeclCXX.cpp:
```diff
@@ -818,6 +818,12 @@ void CodeGenModule::EmitCXXModuleInitFunc(Module *Primary) {
+  if (getTarget().isBranchProtectionSupportedArch(
+          getTarget().getTargetOpts().CPU)) {
+    TargetInfo::BranchProtectionInfo BPI(getLangOpts());
+    getTargetCodeGenInfo().setBranchProtectionFnAttributes(BPI, (*Fn));
```
This does not look like the right place for this code; we have a bunch of different functions for setting attributes, we shouldn't be bypassing all of them to directly juggle branch protection specifically for initializing global variables.

Maybe AArch64TargetCodeGenInfo::setTargetAttributes needs to be changed so it works even if there isn't a FunctionDecl?

vhscampos replied:
Thanks. I'll make the required changes to use `setTargetAttributes`.

ChuanqiXu9 reviewed (commented):
LGTM. I'll leave the formal approval to @efriedma-quic

vhscampos committed:
- Change patch to use existing function.
- Modify setTargetAttributes function to work even without a function
  declaration.

vhscampos committed:
Extend new solution to AArch64

efriedma-quic commented on clang/lib/CodeGen/CGDeclCXX.cpp:
```diff
@@ -793,6 +793,7 @@ void CodeGenModule::EmitCXXModuleInitFunc(Module *Primary) {
     }
     CodeGenFunction(*this).GenerateCXXGlobalInitFunc(Fn, ModuleInits,
                                                      GuardAddr);
+    getTargetCodeGenInfo().setTargetAttributes(nullptr, Fn, *this);
```
This should be inside CreateGlobalInitOrCleanUpFunction .

It looks like for functions with internal linkage, it calls SetInternalFunctionAttributes which calls a bunch of stuff.  But we skip all of that for non-internal functions.  Which is probably wrong.

vhscampos committed:
Change the place where `setTargetAttributes` is called

efriedma-quic commented on clang/test/CodeGenCXX/cxx20-module-initializer-pacbti.cpp:
```diff
@@ -0,0 +1,72 @@
+// RUN: %clang_cc1 -triple thumbv8.1m.main-unknown-none-eabi -emit-module-interface -target-feature +pacbti -mbranch-target-enforce -std=c++20 %s -o %t.pcm
```
I don't think we need all the combinations here in RUN lines; we verify that elsewhere. The important thing is just that the attributes are getting computed at all.

efriedma-quic commented on clang/lib/CodeGen/CGDeclCXX.cpp:
```diff
@@ -447,6 +447,8 @@ llvm::Function *CodeGenModule::CreateGlobalInitOrCleanUpFunction(
   if (Linkage == llvm::GlobalVariable::InternalLinkage)
     SetInternalFunctionAttributes(GlobalDecl(), Fn, FI);
+  else
+    getTargetCodeGenInfo().setTargetAttributes(nullptr, Fn, *this);
```
Do we also need to call SetLLVMFunctionAttributes/SetLLVMFunctionAttributesForDefinition here?

vhscampos committed:
- Call extra functions to initialise function
 - Simplify test on the assumption that other tests already test the
   functionaly of the called functions.

efriedma-quic reviewed (approved):
LGTM
